src/shell/Program.cs

```diff
@@ -1,92 +1,154 @@
-  [Subcommand(typeof(ConnectCommand))]
-  [Subcommand(typeof(ShowCommand))]
-  [Subcommand(typeof(TransferCommand))]
-  class Program
-  {
-    public static async Task<int> Main(string[] args)
+    [Command("neosh", Description = "Neo N3 blockchain shell", UsePagerForHelpText = false)]
+    [Subcommand(typeof(ContractCommand))]
+    [Subcommand(typeof(ConnectCommand))]
+    [Subcommand(typeof(ShowCommand))]
+    [Subcommand(typeof(TransferCommand))]
+    class Program
     {
-      EnableAnsiEscapeSequences();
-
-      var services = new ServiceCollection()
-          .AddSingleton<ExpressChainManagerFactory>()
-          .AddSingleton<TransactionExecutorFactory>()
-          .AddSingleton<IFileSystem, FileSystem>()
-          .AddSingleton<IConsole>(PhysicalConsole.Singleton)
-          .BuildServiceProvider();
-
-      var app = new CommandLineApplication<Program>();
-      app.Option<bool>("--stack-trace", "", CommandOptionType.NoValue, o => o.ShowInHelpText = false, inherited: true);
-      app.Conventions
-          .UseDefaultConventions()
-          .UseConstructorInjection(services);
-
-      try
-      {
-        return await app.ExecuteAsync(args);
-      }
-      catch (CommandParsingException ex)
-      {
-        await Console.Error.WriteLineAsync($"\x1b[1m\x1b[31m\x1b[40m{ex.Message}");
-
-        if (ex is UnrecognizedCommandParsingException uex && uex.NearestMatches.Any())
+        public static async Task<int> Main(string[] args)
         {
-          await Console.Error.WriteLineAsync();
-          await Console.Error.WriteLineAsync("Did you mean this?");
-          await Console.Error.WriteLineAsync("    " + uex.NearestMatches.First());
+            EnableAnsiEscapeSequences();
+
+            var services = new ServiceCollection()
+                .AddSingleton<ExpressChainManagerFactory>()
+                .AddSingleton<TransactionExecutorFactory>()
+                .AddSingleton<IFileSystem, FileSystem>()
+                .AddSingleton<IConsole>(PhysicalConsole.Singleton)
+                .BuildServiceProvider();
+
+            var app = new CommandLineApplication<Program>();
+            app.Option<bool>("--stack-trace", "", CommandOptionType.NoValue, o => o.ShowInHelpText = false, inherited: true);
+            app.Conventions
+                .UseDefaultConventions()
+                .UseConstructorInjection(services);
+            ShellExtensions extensions = LoadShellExtensions(services.GetService<IFileSystem>());
+            try
+            {
+                ShellExtension? extension;
+                if (extensions.TryFindCommand(args, out extension) && extension != null)
+                {
+                    var chainFactory = services.GetRequiredService<ExpressChainManagerFactory>();
+                    var txExecutorFactory = services.GetRequiredService<TransactionExecutorFactory>();
+                    var input = chainFactory != null ? chainFactory.GetConnectionFilePath(string.Empty) : string.Empty;
```

Comment:
GetRequiredService throws if it can't resolve the service, so shouldn't need to check for null here